### PR TITLE
Add --exclude option to auditwheel repair

### LINKED ISSUE #76: Whitelisting of libraries shipped by other wheels
In `turbodbc` we link against `pyarrow` in such a fashion that `turbodbc` picks up the Arrow libraries via a relative RPATH in the same virtualenv. Thus we don't need to include them in the `manylinux1` wheel of Turbodbc. With auditwheel's current behaviour they are sadly grafted into turbodbc's wheel. To fix this, two options come to my mind:

 * whilelist all libraries that are contained in the virtualenv
 * explicitly whitelist some libraries via a commandline option

I'm happy to implement one of these approaches but I would like to hear what the preferred one would be.

### LINKED ISSUE #241: Do not place OpenGL libraries into wheels on manylinux*
These libraries really need to come from the end machine, not the wheel. They don't work without a backing video driver and may not work if the package was built on a machine with a different driver than the end machine (e.g., mixing an nvidia driver `libOpenGL` with a Mesa DRI `libOpenGL`). The backing DRI libraries aren't packaged anyways, so they're of limited utility either way.

### LINKED ISSUE #391: Allow not copying `.so` files provided by other PyPI packages (e.g. `libtorch*.so` from `torch`)
`auditwheels` copies all external `.so` files into `<packagename>.lib` folder. Then packs them into the wheel file. Sometimes the current package (the developer what to audit) only has very small C++ extension, but it references a large `.so` file. For example, the user builds a C++ extension for PyTorch, it references `libtorch*.so` (some files up to **0.5GiB+**). This makes the final wheel file extremely large (~2GiB).

We can add an option that if the `.so` files are located inside `site-packages`, then do not copy them but update the `rpath` of the target.

For example, the user builds a C++ extension for PyTorch, it references `libtorch_python.so`. `auditwheel` can update the `rpath` to:

```text
$ORIGIN/../torch/lib  # the relative path should be calculated carefully
```

where `$ORIGIN/..` is `site-packages`. And then adds `torch` to the install dependency.

### PR DESCRIPTION
**Pull request by martinRenou**
This is a follow-up of #310 by @rossant, only keeping the `exclude` option as suggested by @mayeut

Closes #76
Closes #241
Closes #310
Fixes #391

Original PR description:

----

This is a quick fix to solve an issue I have with my project and for which auditwheel seemed to help. I don't know if this approach is suitable to other users of auditwheel.

I develop a C graphics library ([datoviz](https://github.com/datoviz/datoviz/)) that depends on Vulkan and comes with Cython bindings. https://github.com/datoviz/datoviz/issues/13 for Linux (at least Ubuntu 20.04, that would be a start).

1. As a first step, I compile the C project and get a libdatoviz.so shared library.
2. Then, I build the Cython library, which has a dynamic dependency to libdatoviz.
3. I'd like to bundle both the Cython extension module, and libdatoviz, in the same wheel.
4. I tried to use auditwheel repair. It included dozens of other dependencies in the wheel, including libvulkan and many graphics-related dependent libraries. The compiled wheel installs properly, but it doesn't work properly (there are issues related to windowing, GPU access, etc). Related: https://github.com/pypa/auditwheel/issues/241[](https://github.com/rossant)
5. I made some changes to auditwheel (this pull request) to exclude libvulkan and a few other libraries. Some issues were fixed, but not others.
6. As another approach, I tried to bundle just libdatoviz, and no other library in the wheel. That seems to work, at least on my computer. Chances are that the wheel may not work on other Linux-based operating systems though.
Any help would be appreciated, regarding either this pull request, and/or a way to solve the issue I'm facing. Thanks!

### CODE REVIEW

File: src/auditwheel/main_repair.py
Comment (mayeut) [inline src/auditwheel/main_repair.py:91]: Not sure about the plural of `soname` or using this directly as a noun but it disambiguates wether we're talking file names/file paths or soname.
```suggestion
        help="List of excluded sonames as a comma-separated string",
```

File: src/auditwheel/repair.py
Comment (mayeut) [inline src/auditwheel/repair.py:35]: I think we should have an exact match. Why shouldn't we ?

File: src/auditwheel/repair.py
Comment (mayeut) [inline src/auditwheel/repair.py:41]: It's a one liner used exactly once, having a function does not help readability and/or factorization here.

**Review by mayeut — CHANGES_REQUESTED** [review]
Thanks for your contribution,
I left a couple remarks inlined.
The PR is failing CI:
- we do not have pre-commit.ci setup yet so please make sure to run `nox -s lint` in order to check/repair for linter issues. if you don't know it, [nox](https://nox.thea.codes/en/stable/) is a command-line tool that automates testing in multiple Python environments.
- codecov is complaining of a drop in coverage because there's no test associated with this new feature. Can you please add some tests for this ?

File: src/auditwheel/main_repair.py
Comment (alexsavulescu) [inline src/auditwheel/main_repair.py:89]:
```suggestion
        "--exclude",
```

Comment (pramodk) [reply src/auditwheel/main_repair.py:89]: this is great work @martinRenou ! we need this in hpc environment to exclude libraries like OpenMP/OpenACC.

As Alex suggested, naming option `exclude` is more obvious here? (or, this was just a typo)

Comment (martinRenou) [reply src/auditwheel/main_repair.py:89]: This was a typo indeed, thanks for reviewing!

**Comment by martinRenou** [comment]
Thanks for reviewing! I'll try to find the time to add tests and fix linting

File: src/auditwheel/repair.py
Comment (ferdonline) [inline src/auditwheel/repair.py:53]: I would default this to the empty tuple so that we can simplify logic everywhere it's used
```suggestion
    exclude: List[str] = (),
```

File: src/auditwheel/repair.py
Comment (ferdonline) [reply src/auditwheel/repair.py:35]: This could well be a lambda internal to `repair_wheel`:
```
_exclude_rule = lambda item: next((item in soname for item in exclude), None)
```
But I also wonder about partial string matching. We probably want very good control about whike libraries are excluded.
In case you're thinking about matching groups, then make it explicit, like mylibgroup* (check the glob module)

File: src/auditwheel/repair.py
Comment (martinRenou) [reply src/auditwheel/repair.py]: Thanks!

**Comment by leofang** [comment]
I think this is a nice addition. Any chance we can get this in sooner? 🙂

**Review by auvipy — DISMISSED** [review]
please increase the test coverage

**Comment by martinRenou** [comment]
I don't mind at all. Thanks a lot!

**Comment by lkollar** [comment]
I haven't merged this yet, as I'm still unsure if it is a good idea to exclude arbitrary libraries. On one hand, it solves a real issue, and I don't see how else we could achieve this, but it also ends up violating the manylinux spec. There is a risk that this will cause confusion and bad user experience, and it makes it more likely that users will run into ABI incompatibilities due to version mismatches between the extension and its excluded dependencies. Another concern is that if PyPI wanted to block uploads of non-conforming wheels, this feature would make it more difficult to implement it (see https://github.com/pypi/warehouse/issues/5420).

I wonder if this needs a larger discussion with the PyPA community, since the scope is bigger here than just auditwheel. If we want to properly support wheels which require some special external libraries, that could require changes in tooling in other areas as well. I don't have the bandwidth to take this on, but if someone wanted to raise this on the PyPA Discourse, that would be a good start.

Any thoughts @mayeut?

**Comment by mattip** [comment]
I think this would allow abusing the purpose of auditwheel. I commented on the original issue about not pulling `libOpenGL` shared objects into the wheels.

**Comment by jakirkham** [comment]
> I wonder if this needs a larger discussion with the PyPA community, since the scope is bigger here than just auditwheel. If we want to properly support wheels which require some special external libraries, that could require changes in tooling in other areas as well. I don't have the bandwidth to take this on, but if someone wanted to raise this on the [Packaging Discourse](https://discuss.python.org/c/packaging/14), that would be a good start.

There was [PEP 668]( https://peps.python.org/pep-0668/ ), which AIUI was also trying to help this use case of having some libraries externally managed. This is the result of a few discussions in different threads over the years (including past PyCons):

* https://discuss.python.org/t/drawing-a-line-to-the-scope-of-python-packaging/883
* https://discuss.python.org/t/playing-nice-with-external-package-managers/1968
* https://discuss.python.org/t/pep-668-marking-python-base-environments-as-externally-managed/10302

**Comment by mayeut** [comment]
> I commented on the original issue about not pulling libOpenGL shared objects into the wheels.

This would require a new ABI compatibility analysis to be done to ensure whitelisting is allowed per pep-600

> I think this would allow abusing the purpose of auditwheel.

Yes, it would allow abusing the purpose but allowing package maintainers to use this (responsibly) is a way not to have auditwheel maintainers on the critical path of many projects. As can be seen by the activity in the repo, I do not have that much time to spend on this project, I'm not sure about @lkollar, but I guess that's about the same.

see also my other comment in https://github.com/pypa/auditwheel/pull/310#issuecomment-849773358 

There are already many packages on PyPI abusing PEP600, this is no reason to do so but, IMHO, having an escape hatch is always a good thing to have. It's probably better to have this escape hatch and have the rest of auditwheel run its process than to have a bunch of hand-tagged manylinux1 wheels on PyPI.

One example of abuse (or let's say at least gray area) are GPU dependent packages. Those raised [some discussion](https://discuss.python.org/t/what-to-do-about-gpus-and-the-built-distributions-that-support-them/7125/58) but I'm not sure there's a real conclusion yet (& the discussion starter was package size rather than pep600 compliance).
For example, one of torch wheels is torch-1.12.1-cp310-cp310-manylinux1_x86_64.whl. At least one of the shared object in that wheel links against `libcuda.so.1` which is part of the GPU driver and should not be grafted (nor whitelisted per strict PEP600) thus can't be repaired by auditwheel today (& "this led" to an invalid manylinux1 tag - might be another reason entirely but sure would have been a blocker).

> There is a risk that this will cause confusion and bad user experience, and it makes it more likely that users will run into ABI incompatibilities due to version mismatches between the extension and its excluded dependencies. 

Agreed but my comment above applies: I think the pros outweigh the cons.


> Another concern is that if PyPI wanted to block uploads of non-conforming wheels, this feature would make it more difficult to implement it (see https://github.com/pypi/warehouse/issues/5420).

I think that if PyPI is to implement that feature, they should not block on missing shared objects. They can't know for sure they're not compliant unless doing a manual review. We can't even know for sure that a manylinux1 tagged wheel that embeds some manylinux2014+ only symbol versions is not in fact manylinux1 compliant although in 99.9% of cases, we'd be right to assume it should be tagged manylinux2014.

> I wonder if this needs a larger discussion with the PyPA community, since the scope is bigger here than just auditwheel. If we want to properly support wheels which require some special external libraries, that could require changes in tooling in other areas as well. I don't have the bandwidth to take this on, but if someone wanted to raise this on the [Packaging Discourse](https://discuss.python.org/c/packaging/14), that would be a good start.

While I personally think this should be merged, some discussions should be raised but I don't have the bandwidth to take this on either.

**Comment by leofang** [comment]
> One example of abuse (or let's say at least gray area) are GPU dependent packages. Those raised [some discussion](https://discuss.python.org/t/what-to-do-about-gpus-and-the-built-distributions-that-support-them/7125/58) but I'm not sure there's a real conclusion yet (& the discussion starter was package size rather than pep600 compliance).

Speaking from my personal experience, the whitelisting discussion here is intimately related to the package size discussion. They are the different facets of the same problem.

For GPU packages (or any packages that rely on vender-provided, proprietary prebuilt binaries) without whitelisting they only have the following options:
1. dlopen
2. static linking or auditwheel repair

Option 1 is technically still in violation of the auditwheel spirit, because it's still dynamic linking and could allow a non-self-contained wheel. It also requires a careful runtime check to ensure the correct shared library is dlopen'd. 

Option 2, on the other hand, leads to significantly increased package sizes and, in particular for proprietary packages, potential violation of the vendor's EULA (some vendors put very strict redistribution limitation). Furthermore, to adopt either option requires the downstream package providers to invest significant effort.

I am +1 for whitelisting. The only (minor) disagreement I have with @mayeut's above comment is I see no cons, only pros 🙂

**Comment by mayeut** [comment]
Thanks for your feedback @rgommers.

I edited my comment about a community-wide discussion. I think you're absolutely right on that one. I guess @mattip changed [his view about not wanting this PR to be merged](https://github.com/pypa/auditwheel/pull/368#issuecomment-1217510548).
I will let @lkollar some time to answer your question though.

**Comment by mattip** [comment]
Yes, I changed my view as I learned more about the pain points mentioned above: practicality sometimes beats purity.

**Review by mayeut — APPROVED** [review]
I'll let another week to @lkollar to comment. I'll merge this next week otherwise.

**Comment by XuehaiPan** [comment]
I think this PR also requires an option to update the `rpath` as:

- #392

did. I.e., add the relative path of the skipped libraries in `rpath`. Otherwise, the wheel may be broken for missing libraries.

**Comment by rgommers** [comment]
> I.e., add the relative path of the skipped libraries in `rpath`. Otherwise, the wheel may be broken for missing libraries.

No, please do not do that, it is not going to work. See https://github.com/pypa/auditwheel/issues/391#issuecomment-1272349511 for why, and further down in that discussion for a more robust solution: https://github.com/pypa/auditwheel/issues/391#issuecomment-1272380159.

**Comment by leofang** [comment]
The `rpath` hack does not work for system dependencies (OpenGL, CUDA, Vulkan, ...), which is a wide realm that this PR aims to cover, so please do not do that.

**Comment by mayeut** [comment]
Let's get this merged.
Thanks @martinRenou for the PR.
Thanks to all for your inputs.

**Comment by lkollar** [comment]
Sorry for not answering earlier @mayeut. I was going to agree with merging this though. Thanks!

**Pull request merged**
